dbms/src/Operators/HashJoinProbeTransformOp.cpp

```diff
@@ -55,38 +62,64 @@ OperatorStatus HashJoinProbeTransformOp::onOutput(Block & block)
         {
+        case ProbeStatus::RESTORE_PROBE:
+            // fill probe_process_info for restore probe stage.
+            // The subsequent logic is the same as the probe stage.
+            if (probe_process_info.all_rows_joined_finish)
+            {
+                if (auto ret = probe_transform->tryFillProcessInfoInRestoreProbeStage(probe_process_info); ret != OperatorStatus::HAS_OUTPUT)
+                    return ret;
+            }
         case ProbeStatus::PROBE:
+            // For the probe/restore_probe phase, probe_process_info has been filled,
+            // if all_rows_joined_finish is still true here, it means that there is no input block.
             if unlikely (probe_process_info.all_rows_joined_finish)
             {
-                if (join->finishOneProbe(op_index))
+                if (probe_transform->finishOneProbe())
                 {
-                    // TODO support spill.
-                    RUNTIME_CHECK(!join->hasProbeSideMarkedSpillData(op_index));
-                    join->finalizeProbe();
+                    if (probe_transform->hasMarkedSpillData())
+                    {
+                        switchStatus(ProbeStatus::PROBE_FINAL_SPILL);
+                        BREAK;
+                    }
+                    probe_transform->finalizeProbe();
                 }
-                status = scan_hash_map_after_probe_stream
+                auto next_status = probe_transform->needScanHashMapAfterProbe() || probe_transform->shouldRestore()
                     ? ProbeStatus::WAIT_PROBE_FINISH
                     : ProbeStatus::FINISHED;
-                break;
+                switchStatus(next_status);
+                BREAK;
             }
-            block = join->joinBlock(probe_process_info);
+            block = probe_transform->joinBlock(probe_process_info);
             joined_rows += block.rows();
             return OperatorStatus::HAS_OUTPUT;
+        case ProbeStatus::PROBE_FINAL_SPILL:
+            return OperatorStatus::IO_OUT;
         case ProbeStatus::READ_SCAN_HASH_MAP_DATA:
-            block = scan_hash_map_after_probe_stream->read();
+            block = probe_transform->scanHashMapAfterProbe();
             if unlikely (!block)
             {
-                scan_hash_map_after_probe_stream->readSuffix();
-                status = ProbeStatus::FINISHED;
-                break;
+                probe_transform->endScanHashMapAfterProbe();
+                auto next_status = probe_transform->shouldRestore() ? ProbeStatus::GET_RESTORE_JOIN : ProbeStatus::FINISHED;
+                switchStatus(next_status);
+                BREAK;
             }
             scan_hash_map_rows += block.rows();
             return OperatorStatus::HAS_OUTPUT;
         case ProbeStatus::WAIT_PROBE_FINISH:
-            if (join->isAllProbeFinished())
+            if (probe_transform->quickCheckProbeFinished())
+            {
+                onWaitProbeFinishDone();
+                BREAK;
```

Comment:
I'm not sure if `exec_context.isCancelled()` should be checked unconditionally for all the case branch?

Reply:
Because it is in a while loop, here we check `context.isCancelled` in each loop of while to avoid canceling too slowly.